src/mlpack/core/optimizers/sgd/test_function.cpp

```diff
@@ -35,6 +41,33 @@ double SGDTestFunction::Evaluate(const arma::mat& coordinates, const size_t i)
   }
 }
 
+double SGDTestFunction::Evaluate(const arma::mat& coordinates,
```

Comment:
I think we should remove ``Evaluate(const arma::mat& coordinates, const size_t i)`` and use the batch equivalent. We could also set a default ``batchSize`` of 1. What do you think? Also like to get @rcurtin's input on that.

Reply:
@zoq Yes, that's a good idea :+1: I'll wait until @rcurtin gives his input.

Reply:
Ah, sorry, I forgot to respond to this when I saw it.  I agree, a default `batchSize` of 1 is probably the best way to go here.

Reply:
Okay, what about the ``Gradient`` function, currently it looks like:

```c++
Gradient(const arma::mat& coordinates, const size_t i, arma::mat& gradient)
```

and the version that supports batches looks like:

```c++
Gradient(const arma::mat& coordinates, const size_t begin, const size_t batchSize, arma::mat& gradient)
```

if we reorder the function parameter we can set ``batchSize``to 1 as we would do for the ``Evaluation`` function. Let me know what you think about the idea.

Reply:
Maybe it seems a little awkward, but I might suggest this for reverse compatibility:

```
Gradient(const arma::mat& coordinates, const size_t begin, arma::mat& gradient, const size_t batchSize = 1);
```

I think that's what you meant in your comment, not sure.

Reply:
Okay, I think in the context of reverse compatibility this makes sense, not sure we should release this before mlpack 3.0. But I guess it's nice to have some reverse compatibility between the versions.

Reply:
The idea I wrote also makes it easy to adapt existing optimizers that may not support batch mode, but I don't have a problem with breaking reverse compatibility for mlpack 3.0.0 and using the signature

```
Gradient(const arma::mat& coordinates, const size_t begin, const size_t batchSize, arma::mat& gradient)
```

or whatever else is convenient if you want.  I'd say, pick whatever you think is best here, I don't have too much of an opinion in this case. :)

Reply:
I'll do this: `Gradient(const arma::mat& coordinates, const size_t begin, arma::mat& gradient, const size_t batchSize = 1);`

But, what is wrong with this: `Gradient(const arma::mat& coordinates, const size_t begin, const size_t batchSize = 1, arma::mat& gradient);` ? What does the order in the function signature have to do with reverse compatability?

Reply:
After a parameter with a default argument, all subsequent parameters must have a default argument, or be a parameter pack.